kuma/wiki/tests/test_views.py

```diff
@@ -2549,6 +2550,137 @@ def _check_message_for_headers(message, username):
+        """
+        When a user edits a grandchild document in a watched document tree, we
+        should send an email to users who are watching the tree.
+        """
+        get_current.return_value.domain = 'dev.mo.org'
+
+        root_doc = document(title="Root", slug="Root", save=True)
+        revision(document=root_doc, title="Root", slug="Root", save=True)
+        child_doc = document(title="Child", slug="Child", save=True)
+        child_doc.parent_topic = root_doc
+        child_doc.save()
+        revision(document=child_doc, title="Child", slug="Child", save=True)
+        grandchild_doc = document(title="Grandchild", slug="Grandchild",
+                                  save=True)
+        grandchild_doc.parent_topic = child_doc
+        grandchild_doc.save()
+        revision(document=grandchild_doc, title="Grandchild",
+                 slug="Grandchild", save=True)
+
+        testuser2 = get_user(username='testuser2')
+        EditDocumentInTreeEvent.notify(testuser2, root_doc)
+
+        self.client.login(username='testuser', password='testpass')
+        data = new_document_data()
+        data.update({'form': 'rev',
+                     'slug': grandchild_doc.slug,
+                     'content': 'This edit should send an email',
+                     'comment': 'This edit should send an email'})
+        self.client.post(reverse('wiki.edit',
+                                 args=[grandchild_doc.slug]),
+                         data)
+        eq_(1, len(mail.outbox))
+
+    @mock.patch.object(Site.objects, 'get_current')
+    def test_single_email_when_watching_doc_and_tree(self, get_current):
+        """
+        When a user edits a watched document in a watched document tree, we
+        should only send a single email to users who are watching both the
+        document and the tree.
+        """
+        get_current.return_value.domain = 'dev.mo.org'
+
+        root_doc = document(title="Root", slug="Root", save=True)
+        revision(document=root_doc, title="Root", slug="Root", save=True)
+        child_doc = document(title="Child", slug="Child", save=True)
+        child_doc.parent_topic = root_doc
+        child_doc.save()
+        revision(document=child_doc, title="Child", slug="Child", save=True)
+
+        testuser2 = get_user(username='testuser2')
+        EditDocumentInTreeEvent.notify(testuser2, root_doc)
+        EditDocumentEvent.notify(testuser2, child_doc)
+
+        self.client.login(username='testuser', password='testpass')
+        data = new_document_data()
+        data.update({'form': 'rev',
+                     'slug': child_doc.slug,
+                     'content': 'This edit should send an email',
+                     'comment': 'This edit should send an email'})
+        self.client.post(reverse('wiki.edit',
+                                 args=[child_doc.slug]),
+                         data)
+        eq_(1, len(mail.outbox))
```

Comment:
I think it would be good to check something basic in the emails (the subject?) to also verify it's the email you are expecting.